File: firebase/dart_functions/lib/build_job/retry_handlers.dart
import 'package:firebase_functions/firebase_functions.dart';
import 'package:google_cloud_firestore/google_cloud_firestore.dart';
import 'package:openci_shared/firestore_paths.dart';
import 'package:uuid/uuid.dart';

import '../firebase.dart';
import '../util/github_app.dart';
import '../util/logger.dart';
import '../util/team_auth.dart';

// ---------------------------------------------------------------------------
// Request models
// ---------------------------------------------------------------------------

class RetryBuildJobRequest {
  const RetryBuildJobRequest({required this.buildJobId});

  factory RetryBuildJobRequest.fromJson(Map<String, dynamic> json) {
    final buildJobId = json['buildJobId'] as String?;
    if (buildJobId == null || buildJobId.isEmpty) {
      throw InvalidArgumentError('Missing buildJobId');
    }
    return RetryBuildJobRequest(buildJobId: buildJobId);
  }

  final String buildJobId;
}

class RetryWorkflowRunRequest {
  const RetryWorkflowRunRequest({required this.workflowRunId});

  factory RetryWorkflowRunRequest.fromJson(Map<String, dynamic> json) {
    final workflowRunId = json['workflowRunId'] as String?;
    if (workflowRunId == null || workflowRunId.isEmpty) {
      throw InvalidArgumentError('Missing workflowRunId');
    }
    return RetryWorkflowRunRequest(workflowRunId: workflowRunId);
  }

  final String workflowRunId;
}

// ---------------------------------------------------------------------------
// Handlers
// ---------------------------------------------------------------------------

Future<Map<String, dynamic>> handleRetryBuildJob(
  CallableRequest<RetryBuildJobRequest> request,
  CallableResponse<Map<String, dynamic>> response,
) async {
  final auth = request.auth;
  if (auth == null) {
    throw UnauthenticatedError('Unauthenticated');
  }

  final callerUid = auth.uid;
  final buildJobId = request.data.buildJobId;

  final originalJobRef = firestore
      .collection(buildJobsCollection)
      .doc(buildJobId);
  final originalJobDoc = await originalJobRef.get();

  if (!originalJobDoc.exists) {
    throw NotFoundError('Build job not found');
  }

  final originalJob = originalJobDoc.data()!;

  // Verify team membership
  final teamId = originalJob['teamId'] as String?;
  if (teamId != null) {
    await verifyTeamMembership(auth: auth, teamId: teamId);
  }

  final newDocumentId = const Uuid().v4();
  final checkRunDetailsUrl = buildDashboardRunUrl(newDocumentId);

  // Get fresh GitHub installation token
  final installationId = originalJob['installationId'] as int?;
  String? installationToken;
  String? tokenExpiresAt;
  int? checkRunId = originalJob['checkRunId'] as int?;

  if (installationId != null) {
    try {
      final tokenData = await getInstallationToken(installationId);
      installationToken = tokenData['token'] as String;
      tokenExpiresAt = tokenData['expires_at'] as String;

      final commitSha = originalJob['commitSha'] as String?;
      if (commitSha != null) {
        final workflowName = originalJob['workflowName'] as String?;
        if (workflowName == null) {
          throw Exception('workflowName is missing on job $buildJobId');
        }
        checkRunId = null;
        checkRunId = await createCheckRun(
          token: installationToken,
          owner: originalJob['owner'] as String,
          repo: originalJob['repo'] as String,
          name: workflowName,
          headSha: commitSha,
          status: 'in_progress',
          detailsUrl: checkRunDetailsUrl,
        );
        if (checkRunId != null) {
          logInfo('Created new check run $checkRunId for retry');
        }
      }
    } catch (e) {
      logError('Failed to authenticate with GitHub for retry', null, e);
      throw InternalError('Failed to authenticate with GitHub');
    }
  }

  final now = DateTime.now().toUtc().toIso8601String();

  final newJobData = <String, dynamic>{
    'id': newDocumentId,
    'status': 'queued',
    'owner': originalJob['owner'],
    'repo': originalJob['repo'],
    'teamId': teamId,
    'workflowId': originalJob['workflowId'],
    'workflowFileName': originalJob['workflowFileName'],
    'workflowName': originalJob['workflowName'],
    'jobKey': originalJob['jobKey'],
    'workflowRunId': null,
    'needs': null,
    'resolvedNeeds': null,
    'installationId': installationId,
    'commitSha': originalJob['commitSha'],
    'pullRequestNumber': originalJob['pullRequestNumber'],
    'event': originalJob['event'],
    'action': originalJob['action'],
    'sender': originalJob['sender'],
    'repository': originalJob['repository'],
    'tagName': originalJob['tagName'],
    'installationToken': installationToken,
    'tokenExpiresAt': tokenExpiresAt,
    'checkRunId': checkRunId,
    'runsOn': originalJob['runsOn'],
    'branch': originalJob['branch'],
    'releaseName': originalJob['releaseName'],
    'runCount': 0,
    'latestRunId': null,
    'retriedFromBuildJobId': buildJobId,
    'createdAt': now,
    'updatedAt': now,
  };

  await firestore
      .collection(buildJobsCollection)
      .doc(newDocumentId)
      .set(newJobData);

  logInfo('Build job retried: $buildJobId -> $newDocumentId', {
    'callerUid': callerUid,
    'teamId': teamId,
    'owner': originalJob['owner'],
    'repo': originalJob['repo'],
  });

  return <String, dynamic>{'success': true, 'newBuildJobId': newDocumentId};
}

Future<Map<String, dynamic>> handleRetryWorkflowRun(
  CallableRequest<RetryWorkflowRunRequest> request,
  CallableResponse<Map<String, dynamic>> response,
) async {
  final auth = request.auth;
  if (auth == null) {
    throw UnauthenticatedError('Unauthenticated');
  }

  final callerUid = auth.uid;
  final workflowRunId = request.data.workflowRunId;

  final jobsSnapshot = await firestore
      .collection(buildJobsCollection)
      .where('workflowRunId', WhereFilter.equal, workflowRunId)
      .get();

  if (jobsSnapshot.docs.isEmpty) {
    throw NotFoundError('No jobs found for this workflow run');
  }

  final originalJobs = jobsSnapshot.docs.map((doc) => doc.data()).toList();

  final teamId = originalJobs[0]['teamId'] as String?;
  if (teamId != null) {
    await verifyTeamMembership(auth: auth, teamId: teamId);
  }

  final installationId = originalJobs[0]['installationId'] as int?;
  String? installationToken;
  String? tokenExpiresAt;

  if (installationId != null) {
    try {
      final tokenData = await getInstallationToken(installationId);
      installationToken = tokenData['token'] as String;
      tokenExpiresAt = tokenData['expires_at'] as String;
    } catch (e) {
      logError(
        'Failed to authenticate with GitHub for workflow retry',
        null,
        e,
      );
      throw InternalError('Failed to authenticate with GitHub');
    }
  }

  final newWorkflowRunId = const Uuid().v4();

  // First pass: assign new document IDs for each job
  final newJobDocIds = <String, String>{};
  for (final job in originalJobs) {
    final jobKey = job['jobKey'] as String?;
    if (jobKey != null) {
      newJobDocIds[jobKey] = const Uuid().v4();
    }
  }

  final createdJobIds = <String>[];
  final now = DateTime.now().toUtc().toIso8601String();

  for (final originalJob in originalJobs) {
    final jobKey = originalJob['jobKey'] as String?;
    final newDocumentId = jobKey != null
        ? newJobDocIds[jobKey]!
        : const Uuid().v4();
    final originalNeeds = (originalJob['needs'] as List<dynamic>?)
        ?.cast<String>();
    final hasNeeds = originalNeeds != null && originalNeeds.isNotEmpty;

    Map<String, String>? resolvedNeeds;
    if (hasNeeds) {
      resolvedNeeds = {};
      for (final needKey in originalNeeds) {
        if (newJobDocIds.containsKey(needKey)) {
          resolvedNeeds[needKey] = newJobDocIds[needKey]!;
        } else {
          logWarning(
            'Retry: Job "$jobKey" needs "$needKey" which doesn\'t exist in workflow',
          );
        }
      }
    }

    int? checkRunId;
    if (installationToken != null && originalJob['commitSha'] != null) {
      final workflowName = originalJob['workflowName'] as String?;
      if (workflowName == null) {
        logError('workflowName is missing on job $jobKey');
      } else {
        final multipleJobs = originalJobs.length > 1;
        final checkRunName = multipleJobs
            ? '$workflowName / $jobKey'
            : workflowName;
        final checkRunDetailsUrl = buildDashboardRunUrl(newDocumentId);

        checkRunId = await createCheckRun(
          token: installationToken,
          owner: originalJob['owner'] as String,
          repo: originalJob['repo'] as String,
          name: checkRunName,
          headSha: originalJob['commitSha'] as String,
          status: hasNeeds ? 'queued' : 'in_progress',
          detailsUrl: checkRunDetailsUrl,
        );
        if (checkRunId != null) {
          logInfo('Created new check run $checkRunId for retry $jobKey');
        }
      }
    }

    final newJobData = <String, dynamic>{
      'id': newDocumentId,
      'status': hasNeeds ? 'waiting' : 'queued',
      'owner': originalJob['owner'],
      'repo': originalJob['repo'],
      'teamId': originalJob['teamId'],
      'workflowId': originalJob['workflowId'],
      'workflowFileName': originalJob['workflowFileName'],
      'workflowName': originalJob['workflowName'],
      'jobKey': jobKey,
      'workflowRunId': newWorkflowRunId,
      'needs': originalNeeds,
      'resolvedNeeds': resolvedNeeds,
      'installationId': originalJob['installationId'],
      'commitSha': originalJob['commitSha'],
      'pullRequestNumber': originalJob['pullRequestNumber'],
      'event': originalJob['event'],
      'action': originalJob['action'],
      'sender': originalJob['sender'],
      'repository': originalJob['repository'],
      'tagName': originalJob['tagName'],
      'branch': originalJob['branch'],
      'releaseName': originalJob['releaseName'],
      'installationToken': installationToken,
      'tokenExpiresAt': tokenExpiresAt,
      'checkRunId': checkRunId,
      'runsOn': originalJob['runsOn'],
      'runCount': 0,
      'latestRunId': null,
      'retriedFromWorkflowRunId': workflowRunId,
      'createdAt': now,
      'updatedAt': now,
    };

    await firestore
        .collection(buildJobsCollection)
        .doc(newDocumentId)
        .set(newJobData);
    createdJobIds.add(newDocumentId);
  }

  logInfo(
    'Workflow run retried: $workflowRunId -> $newWorkflowRunId (${createdJobIds.length} jobs)',
    {
      'callerUid': callerUid,
      'teamId': teamId,
      'owner': originalJobs[0]['owner'],
      'repo': originalJobs[0]['repo'],
    },
  );

  return <String, dynamic>{
    'success': true,
    'newWorkflowRunId': newWorkflowRunId,
    'newBuildJobIds': createdJobIds,
  };
}

File: firebase/openci/ts_functions/src/buildJob/retryHandlers.ts
import { randomUUID } from "node:crypto";

import { logger } from "firebase-functions/v2";
import { HttpsError, onCall } from "firebase-functions/v2/https";

import { createCheckRun, getInstallationToken } from "../github/githubApp";
import { buildDashboardRunUrl } from "../github/githubUrls";
import {
  BuildJobStatus,
  createBuildJob,
  getBuildJob as getBuildJobOperation,
  listBuildJobsByWorkflowRun,
} from "../firestoreData";
import { verifyTeamMembership } from "../team/teamAuth";

interface RetryBuildJobRequest {
  buildJobId: string;
}

interface RetryWorkflowRunRequest {
  workflowRunId: string;
}

function requireNonEmptyString(value: unknown, field: string): string {
  if (typeof value !== "string" || value.length === 0) {
    throw new HttpsError("invalid-argument", `${field} is required`);
  }
  return value;
}

async function getBuildJobData(buildJobId: string): Promise<Record<string, unknown>> {
  const result = await getBuildJobOperation({ id: buildJobId });
  if (!result.data.buildJob) {
    throw new HttpsError("not-found", "Build job not found");
  }
  return result.data.buildJob as unknown as Record<string, unknown>;
}

function copyRetryJobFields(
  originalJob: FirebaseFirestore.DocumentData,
  overrides: Record<string, unknown>,
): Record<string, unknown> {
  return {
    owner: originalJob.owner,
    repo: originalJob.repo,
    teamId: originalJob.teamId,
    workflowId: originalJob.workflowId,
    workflowFileName: originalJob.workflowFileName,
    workflowName: originalJob.workflowName,
    jobKey: originalJob.jobKey,
    installationId: originalJob.installationId,
    installationToken: originalJob.installationToken,
    tokenExpiresAt: originalJob.tokenExpiresAt,
    commitSha: originalJob.commitSha,
    pullRequestNumber: originalJob.pullRequestNumber,
    event: originalJob.event,
    action: originalJob.action,
    sender: originalJob.sender,
    repository: originalJob.repository,
    tagName: originalJob.tagName,
    branch: originalJob.branch,
    releaseName: originalJob.releaseName,
    runsOn: originalJob.runsOn,
    githubApiBaseUrl: originalJob.githubApiBaseUrl,
    githubBaseUrl: originalJob.githubBaseUrl,
    ...overrides,
  };
}

function numberFromDataConnectInt64(value: unknown): number | undefined {
  if (typeof value === "number") return value;
  if (typeof value === "string" && value.length > 0) {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : undefined;
  }
  return undefined;
}

function stringFromUnknown(value: unknown): string | undefined {
  return typeof value === "string" && value.length > 0 ? value : undefined;
}

function isInstallationTokenValid(expiresAt: unknown): boolean {
  if (typeof expiresAt !== "string" || expiresAt.length === 0) return false;
  const expiresAtMs = Date.parse(expiresAt);
  if (!Number.isFinite(expiresAtMs)) return false;
  return expiresAtMs > Date.now() + 5 * 60 * 1000;
}

export const retryBuildJob = onCall<
  RetryBuildJobRequest,
  Promise<{ success: true; newBuildJobId: string }>
>(async (request) => {
  const auth = request.auth;
  if (!auth) {
    throw new HttpsError("unauthenticated", "Unauthenticated");
  }

  const buildJobId = requireNonEmptyString(request.data?.buildJobId, "buildJobId");
  const originalJob = await getBuildJobData(buildJobId);
  const teamId = typeof originalJob.teamId === "string" ? originalJob.teamId : undefined;
  if (teamId) {
    await verifyTeamMembership(auth, teamId);
  }

  const newDocumentId = randomUUID();
  const installationId = numberFromDataConnectInt64(originalJob.installationId);
  const apiBaseUrl = stringFromUnknown(originalJob.githubApiBaseUrl);
  let installationToken = stringFromUnknown(originalJob.installationToken);
  let tokenExpiresAt = stringFromUnknown(originalJob.tokenExpiresAt);
  let checkRunId = numberFromDataConnectInt64(originalJob.checkRunId);

  if (
    installationId !== undefined &&
    (!installationToken || !isInstallationTokenValid(tokenExpiresAt))
  ) {
    try {
      const tokenData = await getInstallationToken(installationId, { apiBaseUrl });
      installationToken = tokenData.token;
      tokenExpiresAt = tokenData.expiresAt;
    } catch (error) {
      logger.error("Failed to authenticate with GitHub for retry", { buildJobId, error });
      throw new HttpsError("internal", "Failed to authenticate with GitHub");
    }
  }
  if (installationToken && typeof originalJob.commitSha === "string") {
    const workflowName = requireNonEmptyString(originalJob.workflowName, "workflowName");
    checkRunId =
      (await createCheckRun({
        token: installationToken,
        owner: requireNonEmptyString(originalJob.owner, "owner"),
        repo: requireNonEmptyString(originalJob.repo, "repo"),
        name: workflowName,
        headSha: originalJob.commitSha,
        status: "in_progress",
        detailsUrl: buildDashboardRunUrl(newDocumentId),
        apiBaseUrl,
      })) ?? undefined;
  }

  await createBuildJob(
    copyRetryJobFields(originalJob, {
      id: newDocumentId,
      status: BuildJobStatus.QUEUED,
      workflowRunId: null,
      needs: null,
      resolvedNeeds: null,
      installationToken,
      tokenExpiresAt,
      checkRunId,
      runCount: 0,
      latestRunId: null,
      retriedFromBuildJobId: buildJobId,
    }) as never,
  );

  logger.info("Build job retried", { buildJobId, newDocumentId, callerUid: auth.uid, teamId });
  return { success: true, newBuildJobId: newDocumentId };
});

export const retryWorkflowRun = onCall<
  RetryWorkflowRunRequest,
  Promise<{ success: true; newWorkflowRunId: string; newBuildJobIds: string[] }>
>(async (request) => {
  const auth = request.auth;
  if (!auth) {
    throw new HttpsError("unauthenticated", "Unauthenticated");
  }

  const workflowRunId = requireNonEmptyString(request.data?.workflowRunId, "workflowRunId");
  const jobsResult = await listBuildJobsByWorkflowRun({ workflowRunId });

  if (jobsResult.data.buildJobs.length === 0) {
    throw new HttpsError("not-found", "No jobs found for this workflow run");
  }

  const originalJobs = jobsResult.data.buildJobs;
  const teamId = typeof originalJobs[0]?.teamId === "string" ? originalJobs[0].teamId : undefined;
  if (teamId) {
    await verifyTeamMembership(auth, teamId);
  }

  const installationId = numberFromDataConnectInt64(originalJobs[0]?.installationId);
  const apiBaseUrl = stringFromUnknown(originalJobs[0]?.githubApiBaseUrl);
  let installationToken = stringFromUnknown(originalJobs[0]?.installationToken);
  let tokenExpiresAt = stringFromUnknown(originalJobs[0]?.tokenExpiresAt);
  if (
    installationId !== undefined &&
    (!installationToken || !isInstallationTokenValid(tokenExpiresAt))
  ) {
    try {
      const tokenData = await getInstallationToken(installationId, { apiBaseUrl });
      installationToken = tokenData.token;
      tokenExpiresAt = tokenData.expiresAt;
    } catch (error) {
      logger.error("Failed to authenticate with GitHub for workflow retry", {
        workflowRunId,
        error,
      });
      throw new HttpsError("internal", "Failed to authenticate with GitHub");
    }
  }

  const newWorkflowRunId = randomUUID();
  const newJobDocIds = new Map<string, string>();
  for (const job of originalJobs) {
    if (typeof job.jobKey === "string") {
      newJobDocIds.set(job.jobKey, randomUUID());
    }
  }

  const createdJobIds: string[] = [];

  for (const originalJob of originalJobs) {
    const jobKey = typeof originalJob.jobKey === "string" ? originalJob.jobKey : undefined;
    const newDocumentId = jobKey ? newJobDocIds.get(jobKey)! : randomUUID();
    const originalNeeds = Array.isArray(originalJob.needs)
      ? originalJob.needs.filter((need: unknown): need is string => typeof need === "string")
      : undefined;
    const hasNeeds = Boolean(originalNeeds?.length);
    const resolvedNeeds: Record<string, string> | undefined = hasNeeds ? {} : undefined;
    for (const needKey of originalNeeds ?? []) {
      const newNeedId = newJobDocIds.get(needKey);
      if (newNeedId && resolvedNeeds) {
        resolvedNeeds[needKey] = newNeedId;
      }
    }

    let checkRunId: number | undefined;
    if (installationToken && typeof originalJob.commitSha === "string") {
      const workflowName =
        typeof originalJob.workflowName === "string" ? originalJob.workflowName : undefined;
      if (workflowName) {
        const checkRunName =
          originalJobs.length > 1 && jobKey ? `${workflowName} / ${jobKey}` : workflowName;
        checkRunId =
          (await createCheckRun({
            token: installationToken,
            owner: requireNonEmptyString(originalJob.owner, "owner"),
            repo: requireNonEmptyString(originalJob.repo, "repo"),
            name: checkRunName,
            headSha: originalJob.commitSha,
            status: hasNeeds ? "queued" : "in_progress",
            detailsUrl: buildDashboardRunUrl(newDocumentId),
            apiBaseUrl,
          })) ?? undefined;
      }
    }

    await createBuildJob(
      copyRetryJobFields(originalJob, {
        id: newDocumentId,
        status: hasNeeds ? BuildJobStatus.WAITING : BuildJobStatus.QUEUED,
        workflowRunId: newWorkflowRunId,
        needs: originalNeeds,
        resolvedNeeds,
        installationToken,
        tokenExpiresAt,
        checkRunId,
        runCount: 0,
        latestRunId: null,
        retriedFromWorkflowRunId: workflowRunId,
      }) as never,
    );
    createdJobIds.push(newDocumentId);
  }

  logger.info("Workflow run retried", {
    workflowRunId,
    newWorkflowRunId,
    count: createdJobIds.length,
  });
  return { success: true, newWorkflowRunId, newBuildJobIds: createdJobIds };
});
